src/main/scala/com/normation/cfclerk/domain/AixPasswordHashAlgo.scala

```diff
@@ -0,0 +1,331 @@
+ * they only kepted:
+ * - the number of bytes (and so chars) for the hash: 20 for ssha1, 32 for ssha256, 64 for ssh512
+ * - the base64 encoding table (which is not the standard one but starts with "./012" etc
+ *
+ * What changed is:
+ * - they use PBKDF2 HMAC-(sha1, sha256, sha512) in place of Sha-Crypt,
+ * - they use a different padding table
+ * - the number of iterations, named "rounds" in Unix crypt vocabulary, is not the number N
+ *   found at the begining of the hash string (after the algo name). The number of iteration
+ *   is actually  2^N, and N is called in /etc/security/pwdalg.cfg the "cost"
+ *
+ * Hope this decription may help other people find there way to generate AIX hash string.
+ */
+object AixPasswordHashAlgo extends Loggable {
+  import java.lang.{ StringBuilder => JStringBuilder }
+
+  /**
+   * Table with characters for Sha-Crypt Base64 transformation,
+   * and used by Aix even if they DON'T use sha-crypt.
+   */
+  final val SCB64Table = "./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+
+
+  /**
+   * Sha-Crypt Base64 byte encoding into chars.
+   *
+   * Convert 3 bytes into 0 to 3 chars from the B64Tables.
+   * Output is written into the buffer given in input.
+   *
+   * This code is a Scala adaptation from org.apache.commons.codec.digest.B64
+   *
+   */
+  def b64from24bit(b2: Byte, b1: Byte, b0: Byte, outNumChars: Int, buffer: JStringBuilder) {
+    // The bit masking is necessary because the JVM byte type is signed!
+    var w = ((b2 << 16) & 0x00ffffff) | ((b1 << 8) & 0x00ffff) | (b0 & 0xff)
+    // It's effectively a "for" loop but kept to resemble the original C code.
+    val n = outNumChars
+    for(i <- 0 until n) {
+      buffer.append(SCB64Table.charAt(w & 0x3f))
+      w >>= 6
+    }
+  }
+
+  /**
+   * AIX {smd5} implementation.
+   * @parameter pwd: the string password, in UTF-8, to hash
+   * @parameter salt: a
+   */
+  def smd5(pwd: String, salt: Option[String] = None): String = {
+
+    val s    = salt.getOrElse(getRandomSalt(8))
+    val hash = Md5Crypt.md5Crypt(pwd.getBytes("UTF-8"), s, "")
+
+    val sb = new java.lang.StringBuilder()
+    sb.append("{smd5}")
+    sb.append(hash)
+
+    sb.toString
+  }
+
+
+  /**
+   * Create an AIX {ssha1} hash string from pwd. Optionaly, specify the
+   * salt string to use (default to a random 16 chars string), and the "num_cost"
+   * (defaults to 10, i.e 1024 iterations).
+   *
+   * If PBKDF2WithHmacSHA1 crypto algorithm is not available in the current JVM,
+   * revert back to {smd5}.
+   */
+  def ssha1(pwd: String, salt: Option[String] = None, cost: Int = 10): String = {
+    ssha(pwd, salt, cost, ShaSpec.SHA1).openOr {
+      // this should not happen, because PBKDF2WithHmacSHA1 is
+      // in standard Java since Java6. But who knows..
+      // Fallback to md5 hash.
+      logger.debug("Your current Java installation does not support PBKDF2WithHmacSHA1 algorithm, " +
```

Comment:
I'm not sure. You will end having that message hundreds and thousand of time by generation, and it will fill the logs. Perhaps for md5 it is intersting, because there is a potential security problem (even if not real today). For sha256 -> sha1, I don't think there is a need, there is no foreseen security problem on it.